pkg/puller/puller.go

```diff
@@ -0,0 +1,501 @@
+}
+
+func (p *Puller) recalcPeer(ctx context.Context, peer swarm.Address, po, d uint8) {
+	p.logger.Debugf("puller recalculating peer %s po %d depth %d", peer, po, d)
+	syncCtx := p.syncPeers[po][peer.String()] // recalcPeer is called under lock, this is safe
+
+	syncCtx.Lock()
+	defer syncCtx.Unlock()
+
+	p.cursorsMtx.Lock()
+	c := p.cursors[peer.String()]
+	p.cursorsMtx.Unlock()
+
+	if po >= d {
+		// within depth
+		var want, dontWant []uint8
+
+		for i := d; i < bins; i++ {
+			if i == 0 {
+				continue
+			}
+			want = append(want, i)
+		}
+		for i := uint8(0); i < d; i++ {
+			dontWant = append(dontWant, i)
+		}
+
+		for _, bin := range want {
+			// question: do we want to have the separate cancel funcs per live/hist
+			// for known whether syncing is running on that bin/stream? could be some race here
+			if _, ok := syncCtx.binCancelFuncs[bin]; !ok {
+				// if there's no bin cancel func it means there's no
+				// sync running on this bin. start syncing both hist and live
+				cur := c[bin]
+				binCtx, cancel := context.WithCancel(ctx)
+				syncCtx.binCancelFuncs[bin] = cancel
+				if cur > 0 {
+					go p.histSyncWorker(binCtx, peer, bin, cur)
+				}
+				go p.liveSyncWorker(binCtx, peer, bin, cur)
+			}
+		}
+
+		for _, bin := range dontWant {
+			if c, ok := syncCtx.binCancelFuncs[bin]; ok {
+				// we have sync running on this bin, cancel it
+				c()
+				delete(syncCtx.binCancelFuncs, bin)
+			}
+		}
+	} else {
+		// outside of depth
+		var (
+			want     = po
+			dontWant = []uint8{0} // never want bin 0
+		)
+
+		for i := uint8(0); i < bins; i++ {
+			if i == want {
+				continue
+			}
+			dontWant = append(dontWant, i)
+		}
+
+		if _, ok := syncCtx.binCancelFuncs[want]; !ok {
+			// if there's no bin cancel func it means there's no
+			// sync running on this bin. start syncing both hist and live
+			cur := c[want]
+			binCtx, cancel := context.WithCancel(ctx)
+			syncCtx.binCancelFuncs[po] = cancel
+			if cur > 0 {
+				go p.histSyncWorker(binCtx, peer, want, cur)
+			}
+			go p.liveSyncWorker(binCtx, peer, want, cur)
+		}
+		for _, bin := range dontWant {
+			if c, ok := syncCtx.binCancelFuncs[bin]; ok {
+				// we have sync running on this bin, cancel it
+				c()
+				delete(syncCtx.binCancelFuncs, bin)
+			}
+		}
+	}
+}
+
+func (p *Puller) syncPeer(ctx context.Context, peer swarm.Address, po, d uint8) {
+	syncCtx := p.syncPeers[po][peer.String()] // syncPeer is called under lock, so this is safe
+	syncCtx.Lock()
+	defer syncCtx.Unlock()
+
+	p.cursorsMtx.Lock()
+	c, ok := p.cursors[peer.String()]
+	p.cursorsMtx.Unlock()
+
+	if !ok {
+		cursors, err := p.syncer.GetCursors(ctx, peer)
+		if err != nil {
+			p.logger.Errorf("error getting cursors: %v", err)
+			delete(p.syncPeers[po], peer.String())
+			return
+			// remove from syncing peers list, trigger channel to find some other peer
+			// maybe blacklist for some time
+		}
+		p.cursorsMtx.Lock()
+		p.cursors[peer.String()] = cursors
```

Comment:
i dont understand. you need to keep a copy of the cursors in case you'd like to start syncing on some bins later on (if the depth moves back and forth). you don't want to ask the cursors if you already asked for them before

Reply:
are they stored in db and here you just  cache them after query? then ok

Reply:
no they are only stored in memory (in the map...)